test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java

```diff
@@ -0,0 +1,177 @@
+            page.close();
+        }
+        capacity -= bytesToRelease;
+        internalIndex = Math.max(internalIndex - bytesToRelease, 0);
+        offset = indexInPage(bytesToRelease + offset);
+    }
+
+    public ByteBuffer[] getPreIndexBuffers() {
+        if (internalIndex == 0) {
+            return new ByteBuffer[0];
+        }
+        int indexWithOffset = internalIndex + offset;
+        int pageCount = pageIndex(indexWithOffset);
+        int finalLimit = indexInPage(indexWithOffset);
+        if (finalLimit != 0) {
+            pageCount += 1;
+        }
+
+        ByteBuffer[] buffers = new ByteBuffer[pageCount];
+        Iterator<Page> pageIterator = pages.iterator();
+        ByteBuffer firstBuffer = pageIterator.next().buffer.duplicate();
+        firstBuffer.position(firstBuffer.position() + offset);
+        buffers[0] = firstBuffer;
+        for (int i = 1; i < buffers.length; ++i) {
+            buffers[i] = pageIterator.next().buffer.duplicate();
+        }
+        if (finalLimit != 0) {
+            buffers[buffers.length - 1].limit(finalLimit);
+        }
+
+        return buffers;
+    }
+
+    public ByteBuffer[] getPostIndexBuffers() {
+        if (internalIndex == capacity) {
+            return new ByteBuffer[0];
+        }
+        int indexWithOffset = offset + internalIndex;
+
+        int pageIndex = pageIndex(indexWithOffset);
+        int indexInPage = indexInPage(indexWithOffset);
+
+        ByteBuffer[] buffers = new ByteBuffer[pages.size() - pageIndex];
+        Iterator<Page> pageIterator = pages.descendingIterator();
+        for (int i = buffers.length - 1; i > 0; --i) {
+            buffers[i] = pageIterator.next().buffer.duplicate();
+        }
+        ByteBuffer firstPostIndexBuffer = pageIterator.next().buffer.duplicate();
+        firstPostIndexBuffer.position(firstPostIndexBuffer.position() + indexInPage);
+        buffers[0] = firstPostIndexBuffer;
+
+        return buffers;
+    }
+
+    public void incrementIndex(int delta) {
+        int newIndex = delta + internalIndex;
+        if (newIndex > capacity) {
+            throw new IllegalArgumentException("Cannot increment an index [" + internalIndex + "] with a delta [" + delta +
+                "] that will result in a new index [" + newIndex + "] that is greater than the capacity [" + capacity + "].");
+        }
+        internalIndex = newIndex;
+    }
+
+    public int getIndex() {
+        return internalIndex;
+    }
+
+    public int getCapacity() {
+        return capacity;
+    }
+
+    public int getRemaining() {
+        return capacity - internalIndex;
+    }
+
+    private int numPages(int capacity) {
+        return (capacity + pageMask) >>> pageShift;
+    }
+
+    private int pageIndex(int index) {
+        return index >>> pageShift;
+    }
+
+    private int indexInPage(int index) {
+        return index & pageMask;
+    }
+
+    public static class Page implements Releasable {
```

Comment:
is this a preparation for the future?

Reply:
Yes. As a different `Supplier` would need access. But I'll go ahead and change it to private right now. And can adjust in whenever the "recycling" pr lands.